sdk/go/internal/querybuilder/marshal.go

```diff
@@ -1,18 +1,102 @@
 package querybuilder
 
 import (
+	"context"
 	"fmt"
 	"reflect"
 	"strings"
+	"sync"
+
+	"golang.org/x/sync/errgroup"
+)
+
+// GraphQLMarshaller is an internal interface for marshalling an object into GraphQL.
+type GraphQLMarshaller interface {
+	// XXX_GraphQLType is an internal function. It returns the native GraphQL type name
+	XXX_GraphQLType() string
+	// XXX_GraphqlID is an internal function. It returns the underlying type ID
+	XXX_GraphQLID(ctx context.Context) (string, error)
+}
+
+const (
+	GraphQLMarshallerType = "XXX_GraphQLType"
+	GraphQLMarshallerID   = "XXX_GraphQLID"
 )
 
-func MarshalGQL(v any) string {
-	return marshalGQL(reflect.ValueOf(v))
+var gqlMarshaller reflect.Type
+
+func init() {
+	gqlMarshaller = reflect.TypeOf((*GraphQLMarshaller)(nil)).Elem()
+}
+
+func MarshalGQL(ctx context.Context, v any) (string, error) {
+	vv := reflect.ValueOf(v)
+	preMarshalled, err := resolveCustomMarshallers(ctx, vv)
+	if err != nil {
+		return "", err
+	}
+	return marshalGQL(vv, preMarshalled), nil
+}
+
+func resolveCustomMarshallers(ctx context.Context, v reflect.Value) (map[reflect.Value]string, error) {
+	preMarshalled := map[reflect.Value]string{}
```

Comment:
Using `reflect.Value` as a map key makes me slightly nervous since the docs say that you shouldn't compare `reflect.Value`s (though the exact meaning of that statement is ambiguous): https://pkg.go.dev/reflect#Value

My fear would be that since [there's an `unsafe.Pointer` in the `Value` struct](https://cs.opensource.google/go/go/+/refs/tags/go1.19.2:src/reflect/value.go;l=45) then we could have code that works most of the time but randomly doesn't cause the go runtime moved a value around in memory or weird stuff like that.

This is all completely hypothetical though, easily could be a non-issue. If there's a really easy way to avoid needing to cache this beforehand in this map, then maybe it's worth it. Otherwise maybe just a comment here would be useful in case we hit a weird bug someday and are looking over this code.

Reply:
100% -- I feel really weird about this.

The alternative would be to make a key somehow -- but it gets messy, fast (array, structs, etc).

The simplest way though is to just get rid of this and do things sequentially.

query builder is already parallelizing arguments -- this second parallelization is at the argument level (e.g. run a container array in parallel, for instance the one you added on Publish)

Maybe it's overkill though and I can just remove this? WDYT?

Reply:
Actually, I'll try something else, should be quick.

Reply:
Yeah if it's easy to just remove that extra parallelization for now that seems like the right approach. We can easily re-add it in the future if needed and get a better understanding of the reflect/unsafe stuff.

Reply:
Done!

I found a way to keep the parallelism, get rid of the map key AND simplify the code! It was over engineered to begin with.